File: chw-egg-game/lib/models/egg_model.dart
class EggModel {
  int level;
  int experience;
  int hunger;
  int happiness;
  DateTime lastFeedTime;
  DateTime birthTime;
  
  EggModel({
    this.level = 1,
    this.experience = 0,
    this.hunger = 100,
    this.happiness = 100,
    DateTime? lastFeedTime,
    DateTime? birthTime,
  }) : lastFeedTime = lastFeedTime ?? DateTime.now(),
       birthTime = birthTime ?? DateTime.now();
  
  // 경험치 추가
  void addExperience(int exp) {
    experience += exp;
    while (experience >= getRequiredExp()) {
      levelUp();
    }
  }
  
  // 레벨업에 필요한 경험치
  int getRequiredExp() {
    return level * 100;
  }
  
  // 레벨업
  void levelUp() {
    experience -= getRequiredExp();
    level++;
  }
  
  // 먹이주기
  void feed(int amount) {
    hunger = (hunger + amount).clamp(0, 100);
    happiness = (happiness + 5).clamp(0, 100);
    lastFeedTime = DateTime.now();
  }
  
  // 놀아주기
  void play(int amount) {
    happiness = (happiness + amount).clamp(0, 100);
    hunger = (hunger - 5).clamp(0, 100);
  }
  
  // 시간 경과에 따른 상태 업데이트
  void updateStatus() {
    final now = DateTime.now();
    final hoursPassed = now.difference(lastFeedTime).inHours;
    
    hunger = (hunger - hoursPassed * 5).clamp(0, 100);
    happiness = (happiness - hoursPassed * 3).clamp(0, 100);
  }
  
  // JSON 변환
  Map<String, dynamic> toJson() {
    return {
      'level': level,
      'experience': experience,
      'hunger': hunger,
      'happiness': happiness,
      'lastFeedTime': lastFeedTime.toIso8601String(),
      'birthTime': birthTime.toIso8601String(),
    };
  }
  
  factory EggModel.fromJson(Map<String, dynamic> json) {
    return EggModel(
      level: json['level'],
      experience: json['experience'],
      hunger: json['hunger'],
      happiness: json['happiness'],
      lastFeedTime: DateTime.parse(json['lastFeedTime']),
      birthTime: DateTime.parse(json['birthTime']),
    );
  }
}

File: eggiverse-main/app/src/main/java/com/eggiverse/app/data/EggModel.java
package com.eggiverse.app.data;

import org.json.JSONException;
import org.json.JSONObject;

public class EggModel {
    private int level;
    private int experience;
    private int hunger;
    private int happiness;
    private long lastFeedTime;
    private long birthTime;

    public EggModel() {
        this.level = 1;
        this.experience = 0;
        this.hunger = 100;
        this.happiness = 100;
        long now = System.currentTimeMillis();
        this.lastFeedTime = now;
        this.birthTime = now;
    }

    public EggModel(int level, int experience, int hunger, int happiness, long lastFeedTime, long birthTime) {
        this.level = level;
        this.experience = experience;
        this.hunger = hunger;
        this.happiness = happiness;
        this.lastFeedTime = lastFeedTime;
        this.birthTime = birthTime;
    }

    public void addExperience(int exp) {
        experience += exp;
        while (experience >= getRequiredExp()) {
            experience -= getRequiredExp();
            level++;
        }
    }

    public int getRequiredExp() {
        return level * 100;
    }

    public void feed(int amount) {
        hunger = clamp(hunger + amount);
        happiness = clamp(happiness + 5);
        lastFeedTime = System.currentTimeMillis();
    }

    public void play(int amount) {
        happiness = clamp(happiness + amount);
        hunger = clamp(hunger - 5);
    }

    public void updateStatus() {
        long now = System.currentTimeMillis();
        long hours = Math.max(0, (now - lastFeedTime) / (1000 * 60 * 60));
        hunger = clamp(hunger - (int) (hours * 5));
        happiness = clamp(happiness - (int) (hours * 3));
    }

    private int clamp(int value) {
        return Math.max(0, Math.min(100, value));
    }

    public JSONObject toJson() throws JSONException {
        JSONObject object = new JSONObject();
        object.put("level", level);
        object.put("experience", experience);
        object.put("hunger", hunger);
        object.put("happiness", happiness);
        object.put("lastFeedTime", lastFeedTime);
        object.put("birthTime", birthTime);
        return object;
    }

    public static EggModel fromJson(JSONObject object) {
        return new EggModel(
                object.optInt("level", 1),
                object.optInt("experience", 0),
                object.optInt("hunger", 100),
                object.optInt("happiness", 100),
                object.optLong("lastFeedTime", System.currentTimeMillis()),
                object.optLong("birthTime", System.currentTimeMillis())
        );
    }

    public int getLevel() {
        return level;
    }

    public int getExperience() {
        return experience;
    }

    public int getHunger() {
        return hunger;
    }

    public int getHappiness() {
        return happiness;
    }

    public long getLastFeedTime() {
        return lastFeedTime;
    }

    public long getBirthTime() {
        return birthTime;
    }
}

